### PR TITLE
Add ADC filtering and systick callbacks

### PR DESCRIPTION
**Pull request by StephanCoetzee**
The goal of this change is to enable us to do more sophisticated filtering on our ADCs in order to have a stable force sensing feedback loop while gripping keys before bumping them. This also paves the way for using ADC readings for more than just reporting the raw values. Furthermore, the use of SysTick and SysTick callbacks reaches beyond ADC readings.  

The main changes are:
- Implement a SysTick timer with callbacks
- Add a layer of abstraction between the ADC driver and the functions that need access to the ADC readings
- Add basic moving average filtering to the force ADC readings

**SysTick timer with callbacks:**
In order to do filtering on the ADCs, the ADC channels need to be read at a constant rate. To do so, a timer interrupt can be used. However, the micro-contoller that we are using only has 5 timers and we might want to implement more timed events in the future.

So, instead of dedicating a timer just to periodically read the ADC values, a system tick was implemented. Every millisecond SysTick gets increased. A SysTick callback can be registered. The callback function passed is then called in the specified time.

The SysTick callbacks are checked and called in the protocol main loop in protocol.c and not in the SysTick ISR. Therefore, these callbacks happen in soft realtime.

**ADC abstraction layer**
Before this change, ADC values were read directly from the ADC registers and reported over serial in reports.c. This happened whenever a "|" command was received over serial.

Now, another layer was added to make filtering easier and to allow other functions to access _filtered_ ADC readings from global ADC arrays.

**Filtering**
A basic moving average filter was added to the force ADC readings. Modifying or adding filters in signals.c will be very simple in the future.

### CODE REVIEW

File: main.c
Comment (Jeff-Ciesielski) [inline main.c]: Why a second?

File: signals.c
Comment (Jeff-Ciesielski) [inline signals.c:49]: Why not memcpy?

File: signals.c
Comment (Jeff-Ciesielski) [inline signals.c]: Simple and effective.  I like it.

File: signals.c
Comment (Jeff-Ciesielski) [inline signals.c]: I'm concerned that this might not be fast enough.  We need to get filtered, relatively up-to-date information in between steps in order to force servo correctly.  Ideally we'd get a couple of measurements in between.

File: signals.c
Comment (Jeff-Ciesielski) [inline signals.c:18]: This naming is a little bit confusing.  Perhaps analog_voltage_reading_buffer? (shorter would be better, but I definitely did a double take below where this variable is used.)

File: signals.c
Comment (Jeff-Ciesielski) [inline signals.c]: Indentation seems off here.

File: systick.c
Comment (Jeff-Ciesielski) [inline systick.c]: If possible, could you try to be a little more consistent with spacing?  

Paragraphs of code should have a single line in between.  You do this sometimes, but others you'll have `<code><comment><More code related to comment>` without anything in between.

File: systick.c
Comment (Jeff-Ciesielski) [inline systick.c]: Used camel case here, please stick to snake case for variables.

File: systick.c
Comment (Jeff-Ciesielski) [inline systick.c:67]: I thought bit_true handled the shifting?

File: systick.c
Comment (Jeff-Ciesielski) [inline systick.c]: I'm slightly worried about sorting this here since it could be being called from within a callback.  I think the sorting should be done before execution.

File: systick.c
Comment (Jeff-Ciesielski) [inline systick.c]: This should probably be a while to handle the case where two callbacks occur at the same time. The whole function could be reduced to:

```
while (systick_len && systick_callbacks_array[0].callback_time <= SysTick) {
...
}
```

**Review by Jeff-Ciesielski — CHANGES_REQUESTED** [review]

File: systick.c
Comment (StephanCoetzee) [reply systick.c:67]: Nope.

File: main.c
Comment (StephanCoetzee) [reply main.c]: It was arbitrarily chosen. The SIGNALS_CALLBACK_INTERVAL is set to 20 ms, which is too short in this case - we need to wait for everything to be initialized before scheduling the first callback. A second may be a bit of an overkill.

File: signals.c
Comment (StephanCoetzee) [reply signals.c]: Yes, this value is in milliseconds, so we would be sampling at 50 Hz which is too slow. The sampling frequency depends on how fast we move the gripper while sampling because that will affect how fast the force value will change.

File: signals.c
Comment (StephanCoetzee) [reply signals.c:18]: ~~I stuck with the existing naming, but I agree that it can be improved.~~ The reason that I went with _x is that x is commonly used for the input signal to a filter. When we add more sophisticated filtering, we might need to have an _y buffer too.

File: signals.c
Comment (StephanCoetzee) [reply signals.c]: Thanks - will fix.

File: signals.c
Comment (StephanCoetzee) [reply signals.c:49]: memcpy works too. I didn't think there would be a real performance benefit (I thought memcpy is implemented in a similar way) or that it is very crucial since N_FILTER = 3. But there's no harm in changing to memcpy, thanks!

File: systick.c
Comment (StephanCoetzee) [reply systick.c]: Good point, I'll move it.

File: systick.c
Comment (StephanCoetzee) [reply systick.c]: My reasoning for not implementing a loop is that I do not want to call multiple callbacks in one execution of systick_service_callback and have the rest of the protocol loop wait. I would rather have the next callback being called the next time the protocol loop rolls over. If the callbacks take very long and there are multiple callbacks at the same time, the callbacks will hold back the rest of the system. I would rather have it as a softer real time. Thoughts?

Comment (Jeff-Ciesielski) [reply systick.c]: Hmm.  I have mixed feelings about it. Lets discuss Monday.

File: signals.c
Comment (StephanCoetzee) [reply signals.c]: The sampling rate of the ADC depends on a number of factors. We are also limited as to how fast we can sample.
- The ADC sampling rate depends on the speed of the gripper (i.e. the frequency of the stepper ISR).
- It also depends on the accuracy of the ADC. We only have a 10-bit ADC. If the value of the ADC changes, for example, every 200 steps of the stepper ISR, it is unncessary to sample in between gripper motor steps.
- I agree that in an ideal situation, we would like to get filtered, up-to-date information in between steps to prevent the gripper from overshooting by a few steps. In order to achieve this, we need to sample the ADC at the same frequency or faster than the stepper ISR. However, since our micro-controller does not support interrupt priorities, if we do so we run the risk of having the critical stepper ISR frequently waiting for the non-critical soft-real-time sys_tick ISR which controls the ADC sampling rate.

The goal of this pull request is to implement ADC filtering for a future force feedback loop implementation, but not to have the sampling rate for the force feedback loop sorted out. 

 On another branch I have a force feedback loop which works without filtered ADC data. I suggest that we sort out the ADC sampling rate for the force feedback loop once we implement the force feedback loop after this pull requests gets merged.

I would like to run tests to figure out how fast we should sample for the force feedback loop, but I am waiting for this branch to get reviewed and merged so that I can rebase my force loop branch off of the new 4axis which includes ADC filtering.

File: systick.c
Comment (StephanCoetzee) [reply systick.c]: Just summarizing what we discussed this morning:

If we go with the while loop, we have the benefit of having the callbacks being called closer to real-time, but at the risk of having an infinite loop. The risk of having an infinite loop increases when we implement longer callback functions and if we decrease the interval between system ticks. 

For example, if we change sys_tick to 1 us and you have a function that takes 2us to execute and schedules a callback every 1 us, this callback function will be stuck in the while loop indefinitely.

At this point in time, the benefit of using a while loop is not very apparent, neither is the risk of using it, since we have very few callbacks. It is more of a design strategy if we decide to use the callback loop extensively in the future.

This is certainly up for debate, but I tend to err on the safe side and rather call one callback function per iteration of the protocol main loop.

Comment (Jeff-Ciesielski) [reply systick.c]: Just for the sake of having my side of the discussion in here when we come back to it later:

1uS ticks are a pathological case that wouldn't ever be used in the real world (you need a free-running counter for that sort of precision). If you're hitting infinite loops, this is a design flaw due to spending too much time in the callback functions (which really should be treated as ISRs for all intents and purposes).  

IMO, the benefits are actually apparent, but negligible given our limited use of callbacks at this time.   

As an side note, I'd advise caution in this definition of 'safe side', as in many systems, the timely execution of callbacks in between state transitions is paramount and can lead to bad/broken behavior in many cases if you let it slide.  I understand that this isn't the case for us, but timed events are typically timed for a reason.

Comment (StephanCoetzee) [reply systick.c]: Thanks, that makes sense. 

1 uS was an example - I realize that it is unrealistic, but I am sure there are other more realistic cases with this flaw.

Alternatively, (as mentioned by you off-line) we can count the number of expired callbacks before the loop and only execute the callbacks which expired before the loop had started. By doing so, we avoid the infinite loop scenario, but still have the benefit of a harder real-time.

**Comment by mdicicco** [comment]
@Jeff-Ciesielski did you finish looking over this?  I can merge if we're ready.

**Comment by Jeff-Ciesielski** [comment]
@mdicicco: I did, however @StephanCoetzee mentioned that he wanted to potentially deprecate this PR in favor of another one. 

Stephan: If this is going away, could you close it for now?

**Comment by StephanCoetzee** [comment]
I will make a new PR shortly which includes this functionality and the force servoing functionality. 

I had to test force sensing with filtering so I merged this branch into a new branch since didn't want to wait for this branch to get merged before continuing with gripping. At this stage it is simpler to close this PR and reference the discussion in this PR in the new PR.

**Comment by StephanCoetzee** [comment]
I will make a new PR shortly which includes this functionality and the force servoing functionality.

I had to test force sensing with filtering so I merged this branch into a new branch since I didn't want to wait for this branch to get merged before continuing with gripping. At this stage it is simpler to close this PR and reference the discussion in this PR in the new PR.

**Pull request closed without merging**
